src/algebraic_geometry/elliptic_curve/weierstrass.lean

```diff
@@ -378,17 +382,119 @@ See Zulip thread:
   is_domain W.coordinate_ring :=
 by { classical, apply_instance }
 
-/-- The function field $R(W) := \mathrm{Frac}(R[W])$ of `W`. -/
-@[reducible] def function_field : Type u := fraction_ring W.coordinate_ring
+variables (x : R) (y : R[X])
+
+/-- The class of the element $X - x$ in $R[W]$ for some $x \in R$. -/
+@[simp] noncomputable def X_class : W.coordinate_ring := adjoin_root.mk W.polynomial $ C $ X - C x
+
+lemma X_class_ne_zero [nontrivial R] : X_class W x ≠ 0 :=
+begin
+  intro hx,
+  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hx) with p hp,
+  apply_fun degree at hp,
+  rw [W.monic_polynomial.degree_mul, degree_polynomial, degree_C $ X_sub_C_ne_zero x] at hp,
+  cases p.degree; cases hp
+end
+
+/-- The class of the element $Y - y(X)$ in $R[W]$ for some $y(X) \in R[X]$. -/
+@[simp] noncomputable def Y_class : W.coordinate_ring := adjoin_root.mk W.polynomial $ X - C y
+
+lemma Y_class_ne_zero [nontrivial R] : Y_class W y ≠ 0 :=
+begin
+  intro hy,
+  cases ideal.mem_span_singleton'.mp (ideal.quotient.eq_zero_iff_mem.mp hy) with p hp,
+  apply_fun degree at hp,
+  rw [W.monic_polynomial.degree_mul, degree_polynomial, degree_X_sub_C] at hp,
+  cases p.degree; cases hp
+end
+
+/-- The ideal $\langle X - x \rangle$ of $R[W]$ for some $x \in R$. -/
+@[simp] noncomputable def X_ideal : ideal W.coordinate_ring := ideal.span {X_class W x}
+
+/-- The ideal $\langle Y - y(X) \rangle$ of $R[W]$ for some $y(X) \in R[X]$. -/
+@[simp] noncomputable def Y_ideal : ideal W.coordinate_ring := ideal.span {Y_class W y}
+
+variable {W}
+
+lemma smul (p : R[X]) (q : W.coordinate_ring) : p • q = adjoin_root.mk W.polynomial (C p) * q :=
+(algebra_map_smul W.coordinate_ring p q).symm
+
+lemma smul_basis_eq_zero [nontrivial R] {p q : R[X]}
+  (hpq : p • 1 + q • adjoin_root.mk W.polynomial X = 0) : p = 0 ∧ q = 0 :=
+begin
+  have h := fintype.linear_independent_iff.mp (basis W).linear_independent ![p, q],
+  erw [fin.sum_univ_succ, basis_zero, fin.sum_univ_one, basis_one] at h,
+  exact ⟨h hpq 0, h hpq 1⟩
+end
+
+lemma exists_smul_basis_eq [nontrivial R] (p : W.coordinate_ring) :
+  ∃ r q : R[X], r • 1 + q • adjoin_root.mk W.polynomial X = p :=
+begin
+  have h := (basis W).mem_span p,
+  rwa [coe_basis, matrix.range_cons_cons_empty, submodule.mem_span_pair] at h
+end
+
+variable (W)
+
+lemma norm_smul_basis [nontrivial R] (p q : R[X]) :
+  algebra.norm R[X] (p • 1 + q • adjoin_root.mk W.polynomial X)
+    = p ^ 2 - p * q * (C W.a₁ * X + C W.a₃)
+      - q ^ 2 * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆) :=
+begin
+  have h : (p • 1 + q • adjoin_root.mk W.polynomial X) * adjoin_root.mk W.polynomial X
+    = (q * (X ^ 3 + C W.a₂ * X ^ 2 + C W.a₄ * X + C W.a₆)) • 1
+      + (p - q * (C W.a₁ * X + C W.a₃)) • adjoin_root.mk W.polynomial X :=
+  begin
+    simp only [smul],
+    exact adjoin_root.mk_eq_mk.mpr
+      ⟨C q, by { simp only [weierstrass_curve.polynomial, C_sub, C_mul], ring1 }⟩
+  end,
```

Comment:
seems you've forgotten the usual style :)
I might come back later to extract a lemma about the matrix of the generator of a power basis, and of `adjoin_root.mk X` the root of a monic polynomial.
```suggestion
      + (p - q * (C W.a₁ * X + C W.a₃)) • adjoin_root.mk W.polynomial X,
  { simp only [smul],
    exact adjoin_root.mk_eq_mk.mpr
      ⟨C q, by { simp only [weierstrass_curve.polynomial, C_sub, C_mul], ring1 }⟩ },
```

Reply:
What's preferable? I've always used `begin ... end` for `have` blocks so I genuinely didn't know about this.